test/e2e/automigration/automigration.go

```diff
@@ -228,4 +238,343 @@ var _ = ginkgo.Describe("auto migration", autoMigrationTestLabel, func() {
 			}
 		})
 	})
+
+	ginkgo.It("Should migrate unschedulable pods properly with overlapping label selectors", func(ctx context.Context) {
+		// 1. Get test clusters
+
+		sort.Slice(clusters, func(i, j int) bool {
+			return clusters[i].Name < clusters[j].Name
+		})
+		clusterToMigrateFromForDp1 := clusters[0]
+		clusterToMigrateFromForDp2 := clusters[2]
+
+		var err error
+
+		// 2. Initialize replicas
+
+		replicasPerCluster := int32(3)
+		totalReplicas := int32(len(clusters)) * replicasPerCluster
+
+		// 3. Initialize deployments
+
+		dp1 := resources.GetSimpleDeployment(f.Name())
+		dp1.Spec.Replicas = pointer.Int32(totalReplicas)
+		maxSurge := intstr.FromInt(0)
+		dp1.Spec.Strategy = appsv1.DeploymentStrategy{
+			Type: appsv1.RollingUpdateDeploymentStrategyType,
+			RollingUpdate: &appsv1.RollingUpdateDeployment{
+				// Prevent upgrade getting stuck
+				MaxSurge: &maxSurge,
+			},
+		}
+		dp2 := dp1.DeepCopy()
+
+		// 4. Create policies
+
+		propPolicy.Spec.AutoMigration.KeepUnschedulableReplicas = true
+		ginkgo.By("Creating PropagationPolicy", func() {
+			propPolicy, err = f.HostFedClient().CoreV1alpha1().PropagationPolicies(f.TestNamespace().Name).Create(
+				ctx,
+				propPolicy,
+				metav1.CreateOptions{},
+			)
+			gomega.Expect(err).ToNot(gomega.HaveOccurred(), framework.MessageUnexpectedError)
+		})
+
+		// This function creates an override policy that results in only a single pod of a deployment being schedulable
+		// in a given cluster using a combination of node affinity and pod anti-affinity.
+		getOverridePolicy := func(cluster *fedcorev1a1.FederatedCluster, deploy *appsv1.Deployment) *fedcorev1a1.OverridePolicy {
+			var selectedNode string
+			var selectedLabelKey, selectedLabelValue string
+
+			for k, v := range dp1.Spec.Template.GetLabels() {
+				selectedLabelKey = k
+				selectedLabelValue = v
+				break
+			}
+
+			nodeList, err := f.ClusterKubeClient(ctx, clusterToMigrateFromForDp1).CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+			gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+			nodes := nodeList.Items
+			gomega.Expect(nodes).ShouldNot(gomega.BeEmpty())
+
+			selectedNode = nodes[0].Name
+			return policies.OverridePolicyForClustersWithPatches(f.Name(), map[string]fedcorev1a1.Overriders{
+				cluster.Name: {
+					JsonPatch: []fedcorev1a1.JsonPatchOverrider{
+						{
+							Operator: "add",
+							Path:     "/spec/template/spec/affinity",
+							Value: apiextensionsv1.JSON{
+								Raw: []byte(
+									//nolint:lll
+									`{"nodeAffinity": {"requiredDuringSchedulingIgnoredDuringExecution": {"nodeSelectorTerms": []}}, "podAntiAffinity": {"requiredDuringSchedulingIgnoredDuringExecution": []}}`,
+								),
+							},
+						},
+						{
+							Path: "/spec/template/spec/affinity/nodeAffinity/requiredDuringSchedulingIgnoredDuringExecution/nodeSelectorTerms",
```

Comment:
With PodAntiAffinity, is NodeAffinity necessary?

Reply:
yes, without nodeaffinity all the nodes will need to have the same topology key